### ISSUE TITLE
Timeout up vs. down the call stack

### ISSUE THREAD

**Issue by woutdenolf** [gevent/gevent#1861]
What is the best way to do this?

```python
try:
    ...  # code that might wait for events or greenlets and therefore could raise Timeout
except gevent.Timeout:
   ...  # cleanup
   if XXXX:
      raise  # caused by timeout context up the call stack
```

In other words, if the timeout is caused by a context up the stack, it needs to bubble up, otherwise it needs to be absorbed (or whatever the logic is in the situation at hand).

What should `XXXX` be?

**Comment by woutdenolf** [comment]
The more I think about it, the more I don't understand why `Timeout` is raise explicitly inside parts of the gevent code base (for example [gevent.event.Event.get](https://github.com/gevent/gevent/blob/5858d226045f701deb8e4b49c03db4cab63053d6/src/gevent/event.py#L351)).

To me it seems `Timeout` should be thought of as an interrupt:
* `gevent.Timeout`: interrupt of the current greenlet, triggered from inside the greenlet (up the call stack)
* `gevent.GreenletExit`: interrupt of the current greenlet, triggered from outside the greenlet
* `KeyboardInterrupt`: interrupt of the process, triggered from outside the process

For example this gevent code:

```python
try:
    someevent.get(timeout=1)
except gevent.Timeout:
    ... # cleanup code
    if <if_does_not_come_event_get>:
        raise
```

would be like this python code

```python
def somefunc():
    raise KeyboardInterrupt

try:
    somefunc()
except KeyboardInterrupt:
    ... # cleanup code
    if <if_does_not_come_from_somefunc>:
        raise
```

That's an inproper usage of `KeyboardInterrupt`. If you raise it you really want to interrupt the process. So this is the proper usage:

```python
def somefunc():
    raise KeyboardInterrupt

try:
    somefunc()
except KeyboardInterrupt:
    ... # cleanup code
    raise  # ALWAYS re-raise
```

But you cannot do that for `Timeout`

```python
try:
    someevent.get(timeout=1)
except gevent.Timeout:
    ... # cleanup code
    raise  # cannot ALWAYS re-raise because `someevent.get` raises
           # it without the intention to interrupt the current greenlet
           # (it would have raised `GreenletExit` otherwise)
```

In other words, `Timeout` should never be raised explicitly. Either raise `TimeoutError` (the call times out) or `GreenletExit` (someone up or down the call stack want to interrupt the current greenlet). For example `Event.get` should clearly raise `TimeoutError`. But it raises `Timeout` so I must be missing the concept of `Timeout`.

**Comment by woutdenolf** [comment]
I'm well aware of the ["catching timeouts docs"](http://www.gevent.org/api/gevent.timeout.html#gevent.Timeout). That's the only safe usage of `gevent.Timeout` I can think of. For example this is not safe

```python
try:
    someevent.get(timeout=1)
except gevent.Timeout:
    ... # cleanup code
```
while this is safe
```python
try:
    with gevent.Timeout(1) as local_timeout:
        someevent.get()
except gevent.Timeout as raised_timeout:
    if local_timeout is raised_timeout:
        ... # cleanup code
    else:
        raise
```
Also this is not safe (`TimeoutError` could be raised and captured anywhere down the call stack):
```python
with gevent.Timeout(1, TimeoutError):
     ...
```
while this is safe
```python
try:
    with gevent.Timeout(1) as local_timeout:
        ...
except gevent.Timeout as raised_timeout:
    if local_timeout is raised_timeout:
        raise TimeoutError
    else:
        raise
```
So my conclusion seems to be:
1. the ["catching timeouts pattern"](http://www.gevent.org/api/gevent.timeout.html#gevent.Timeout) is the only safe usage of `gevent.Timeout`
2. never use any gevent call that could raise `gevent.Timeout` (!!!!!)
3. never capture `gevent.Timeout` unless you're handling a specific instance to which you have a reference
4. never pass an exception to the `gevent.Timeout` constructor
